contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceMode ipc = 3;
+	// User namespace for this container/sandbox.
+	// Note: There is currently no way to set CONTAINER scoped user namespace in the
+	// Kubernetes API. The container runtime has to run with user namespace enabled,
+	// and kubelet then forces the use of host user namespace.
+	// Namespaces currently set by the kubelet: POD, NODE
+	NamespaceModeValue user = 4;
+}
+
+// NamespaceModeValue is the wrapper of NamespaceMode, to allow non-POD default.
+message NamespaceModeValue {
+	NamespaceMode value = 1;
+}
+```
+
+### Runtime Support
+- Docker: https://docs.docker.com/engine/security/userns-remap/
+    - Concerns:
+Docker API does not provide user-namespace mapping. Therefore to handle `RuntimeConfigInfo` API, changes will be done in `dockershim` to read system files, `/etc/subuid` and `/etc/subgid`, for figuring out default user-namespace mapping. It won’t be possible to use custom user/group name to `--userns-remap parameter`, the implementation will expect `dockremap`, which is the default username.
+- CRI-O: WIP https://github.com/nalind/cri-o/tree/idmapping
+
+### Implementation Roadmap
+#### Phase 1: Support in Kubelet, Alpha, [Target: Kubernetes v1.11]
+- Add feature gate `NodeUserNamespace`, disabled by default
+- Add new CRI API, `RuntimeConfigInfo()`
+- Add logic in Kubelet to handle pod creation which includes parsing RuntimeConfigInfo response and changing file-permissions in /var/lib/kubelet with learned userns mapping.
+- Add changes in dockershim to implement RuntimeConfigInfo() for docker runtime
+- Add changes in CRI-O to implement userns support and RuntimeConfigInfo() support
+- Unit test cases
+
+#### Phase 2: Beta Support [Target: Kubernetes v1.12]
+- PSP integration
+- To grow ExperimentalHostUserNamespaceDefaulting from experimental feature gate to a Kubelet flag
+- API changes to allow pod able to request HostUserNamespace in pod spec
+- e2e tests
+
+### References
+- Default host user namespace via experimental flag
+  - https://github.com/kubernetes/kubernetes/pull/31169
+- Enable userns support for containers launched by kubelet
+  - https://github.com/kubernetes/features/issues/127
+- Track Linux User Namespaces in the Pod Security Policy
+  - https://github.com/kubernetes/kubernetes/issues/59152
+- Add support for experimental-userns-remap-root-uid and experimental-userns-remap-root-gid options to match the remapping used by the container runtime.
+  - https://github.com/kubernetes/kubernetes/pull/55707
+- rkt User Namespaces Background
+  - https://coreos.com/rkt/docs/latest/devel/user-namespaces.html
+
+## Future Work
+- Named User namespaces: There is no runtime today which supports creating containers with a named user namespace. For example here is the discussion related to this support in Docker https://github.com/moby/moby/issues/28593 and support is being created for CRI-O https://github.com/nalind/cri-o/tree/idmapping.
```

Comment:
Link to the cri-o PR instead of the fork's branch: https://github.com/kubernetes-incubator/cri-o/pull/1519

Reply:
Can someone please explain what a "named user namespace" means?